### PR TITLE
recall loss

### PR DESCRIPTION
**Pull request by yHarai**
small classの精度向上のためのLossを提案

各クラスのRecallによって動的に重みをつけたCross Entropy Loss（Recall Loss）を提案
(1-Recall)=False Negative Rateで重み付けしている
 
SynthiaデータセットにおいてRecall LossではCE LossやFocal Lossと比べてaccuracy, iouを上回った
 
文献：https://arxiv.org/pdf/2106.14917.pdf
コード：https://github.com/PotatoTian/recall-semseg/blob/main/ptsemseg/loss/recall_loss.py

### CODE REVIEW

**Review by Yuto-web — COMMENTED** [review]
LGTMです！

**Review by Yuto-web — COMMENTED** [review]
class weightsは適用しないんですか？（LGTMと言っておきながらすみません...）

**Comment by yHarai** [comment]
@Yuto-web 
> class weightsは適用しないんですか？（LGTMと言っておきながらすみません...）

recall lossのポイントはclass weightsが(1-recall)になるところだと思うので、あえて適用しませんでした。
必要あれば適用しても良いと思います。

File: farmer/ncc/losses/functional.py [hunk 1]
```diff
@@ -246,3 +246,14 @@ def active_contour_loss(gt, pr, w_region=1.0, w_region_in=1.0, w_region_out=1.0)
...
     return loss
+
+def recall_loss(gt, pr):
+    tp, fp, fn = _tp_fp_fn(gt, pr)
+    recall = tp / (fn + tp)
+    
+    pr = tf.clip_by_value(pr, SMOOTH, 1 - SMOOTH)
+    ce = -gt * K.log(pr)
+    
+    loss = - (1 - recall) * ce
```
Comment (yumion) [inline farmer/ncc/losses/functional.py]: 論文のEq(5)のマイナスはCEのものですので、ここのマイナスが余分だと思います
image

**Comment by yumion** [comment]
@yHarai 

> recall lossのポイントはclass weightsが(1-recall)になるところだと思うので、あえて適用しませんでした。
必要あれば適用しても良いと思います。

focal lossでも`alpha`というパラメータでclass_weight適用しているので、実装してもいいと思います。

**Comment by aiorhiroki** [comment]
良いと思います！
まだmmsegじゃなくてfarmer使うんですね笑

File: farmer/ncc/losses/functional.py (hunk 1 above)
Comment (yHarai) [reply farmer/ncc/losses/functional.py]: ご指摘ありがとうございます。

Comment (yumion) [reply farmer/ncc/losses/functional.py]: lossの結果ってマイナスにならなかったですか？

Comment (yHarai) [reply farmer/ncc/losses/functional.py]: floodingの計算で正になっていました。
`    return tf.math.abs(loss - b) + b`
https://github.com/aiorhiroki/farmer.tf2/blob/develop/farmer/ncc/losses/functional.py#L57
学習結果的には間違っていませんでしたが、コード上は間違っているので修正いたします。
